### DISCUSSION TITLE
Opportunities for improving performance of factorial and prime testing

### DISCUSSION THREAD

**Discussion by gwhitney** [josdejong/mathjs#2786] [Ideas]
This discussion serves as a pointer to PR #1837, which has been closed due to its age and lack of current effort. However, the ideas and the discussion there may still be valid improvements to mathjs. The discussion can be continued here, and we can put pointers to any future PR that may implement some or all of the ideas of that older PR.

**Comment by bartekleon** [comment]
i have found google's repository which makes "Bigint" entirely with javascript and is proposal compilant
https://github.com/GoogleChromeLabs/jsbi
Since the mathjs should work with older versions of JS, maybe that library could be used to polyfill Bigint? It works internally on base 2 instead of base 10, so it should be kept in mind if conversion between that and bigdecimal is needed. It could probably help speeding up algorithms which require big numbers. The library could probably be tweaked as well for max performance for particular cases.

**Comment by bartekleon** [comment]
I was also thinking if it is possible to override mathjs functions. Especially ones using big integers. Like, there could be a plugin for mathjs which overrides the functions that could benefit from using bigint. I would see it like
```html
<script src="./mathjs.js"></script> <!-- basic functionality -->
<script src="./mathjs-bigint.js"></script> <!-- extended functionality if user knows bigint will be supported -->
```

Also to answer possible questions about "but we can still test for Bigint and use contructor" -- yes. But it is orders of magnitude slower to use constructor instead of numeral, so much that it is just not worth it. The solution might be to cache all commonly used values, but i dont think it is the best solution (it sure is messy tho)

**Comment by gwhitney** [comment]
I think all of the supported browsers have native bigint now (since IE11 was dropped), and so I think bigint support is planned as soon as there is a go/no-go decision on a possible TypeScript rewrite (see #2742). Certainly I need it for the project that got me involved with mathjs in the first place.

**Comment by josdejong** [comment]
Yes, `bigint` support will be a great addition. See #2737. 

For the time being it is fine to use `bigint` internally. But having full, public support will be much more work.
